File: packages/dsp-pipeline/src/constants/signal_bands.py
"""Canonical EEG frequency band definitions — mirrors signal_bands.ts exactly."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

BandName = Literal["delta", "theta", "alpha", "beta", "gamma", "high_gamma"]


@dataclass(frozen=True)
class BandDefinition:
    name: BandName
    low_hz: float
    high_hz: float
    description: str


SIGNAL_BANDS: dict[BandName, BandDefinition] = {
    "delta": BandDefinition(
        name="delta",
        low_hz=0.5,
        high_hz=4.0,
        description="Deep sleep, slow cortical potentials (SCP). Used in SCP-BCI paradigms.",
    ),
    "theta": BandDefinition(
        name="theta",
        low_hz=4.0,
        high_hz=8.0,
        description="Drowsiness, memory encoding, frontal midline activity.",
    ),
    "alpha": BandDefinition(
        name="alpha",
        low_hz=8.0,
        high_hz=12.0,
        description=(
            "Sensorimotor idle rhythm (mu). ERD during motor imagery = left/right control."
        ),
    ),
    "beta": BandDefinition(
        name="beta",
        low_hz=18.0,
        high_hz=26.0,
        description="Active motor control, cognitive engagement. ERS after movement cessation.",
    ),
    "gamma": BandDefinition(
        name="gamma",
        low_hz=30.0,
        high_hz=80.0,
        description="High-level cognitive binding, attention. Sensitive to muscle artifact.",
    ),
    "high_gamma": BandDefinition(
        name="high_gamma",
        low_hz=80.0,
        high_hz=200.0,
        description=(
            "Broadband high-gamma: primary cortex activity in ECoG. "
            "Not reliably in scalp EEG."
        ),
    ),
}

MOTOR_IMAGERY_BANDS: list[BandName] = ["alpha", "beta"]
P300_BANDS: list[BandName] = ["delta", "theta"]
EMG_CONTAMINATED_BANDS: list[BandName] = ["gamma", "high_gamma"]

BAND_NAMES: list[BandName] = list(SIGNAL_BANDS.keys())


def band_ranges() -> dict[str, tuple[float, float]]:
    """Return {band_name: (low_hz, high_hz)} for spectral integration."""
    return {name: (band.low_hz, band.high_hz) for name, band in SIGNAL_BANDS.items()}

File: jobs/00_shared_contracts/constants/signal_bands.ts
/**
 * Canonical EEG frequency band definitions.
 *
 * Used by the DSP pipeline for spectral estimation and by the Intent Engine
 * for feature selection. All agents must use these constants — never hardcode
 * frequency ranges inline.
 *
 * Reference: standard clinical and BCI literature band definitions,
 * consistent with BCI2000 implementations (Schalk et al., 2004).
 */

import type { BandName } from "../schema/FeatureVector.js";

export interface BandDefinition {
  /** Band name, matches the BandName union type. */
  name: BandName;
  /** Lower cutoff frequency in Hz (inclusive). */
  lowHz: number;
  /** Upper cutoff frequency in Hz (inclusive). */
  highHz: number;
  /** Typical BCI relevance — what this band encodes. */
  description: string;
}

/**
 * Ordered list of all canonical EEG bands, from lowest to highest frequency.
 */
export const SIGNAL_BANDS: Record<BandName, BandDefinition> = {
  delta: {
    name: "delta",
    lowHz: 0.5,
    highHz: 4,
    description: "Deep sleep, slow cortical potentials (SCP). Used in SCP-BCI paradigms.",
  },
  theta: {
    name: "theta",
    lowHz: 4,
    highHz: 8,
    description: "Drowsiness, memory encoding, frontal midline activity.",
  },
  alpha: {
    name: "alpha",
    lowHz: 8,
    highHz: 12,
    description: "Sensorimotor idle rhythm (mu). ERD during motor imagery = left/right control.",
  },
  beta: {
    name: "beta",
    lowHz: 18,
    highHz: 26,
    description: "Active motor control, cognitive engagement. ERS after movement cessation.",
  },
  gamma: {
    name: "gamma",
    lowHz: 30,
    highHz: 80,
    description: "High-level cognitive binding, attention. Sensitive to muscle artifact.",
  },
  high_gamma: {
    name: "high_gamma",
    lowHz: 80,
    highHz: 200,
    description: "Broadband high-gamma: primary cortex activity in ECoG. Not reliably in scalp EEG.",
  },
};

/**
 * The two bands most critical for motor imagery BCI classification.
 * BCI2000 cursor control uses power in these bands at sensorimotor cortex (C3/C4).
 */
export const MOTOR_IMAGERY_BANDS: BandName[] = ["alpha", "beta"];

/**
 * Bands used for P300 evoked potential detection.
 * P300 is a broadband response peaking ~300 ms post-stimulus, prominent in delta/theta.
 */
export const P300_BANDS: BandName[] = ["delta", "theta"];

/**
 * Bands most susceptible to EMG artifact contamination.
 * Artifact rejection should be applied before computing these bands.
 */
export const EMG_CONTAMINATED_BANDS: BandName[] = ["gamma", "high_gamma"];
